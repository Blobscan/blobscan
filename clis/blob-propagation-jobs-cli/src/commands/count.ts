import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import type { QueueHumanName } from "../Context";
import { context } from "../context-instance";
import type { Command } from "../types";
import { helpOptionDef, queuesOptionDef } from "../utils";

const countCommandOptDefs: commandLineArgs.OptionDefinition[] = [
  helpOptionDef,
  {
    ...queuesOptionDef,
    description: `Queue to get the jobs count of. ${queuesOptionDef.description}`,
  },
];

export const countCommandUsage = commandLineUsage([
  {
    header: "Count Command",
    content: "Count propagation jobs.",
  },
  {
    header: "Options",
    optionList: countCommandOptDefs,
  },
]);

export const count: Command = async function (argv) {
  const { help, queue: queueNames } = commandLineArgs(countCommandOptDefs, {
    argv,
  }) as {
    help?: boolean;
    queue?: QueueHumanName[];
  };

  if (help) {
    console.log(countCommandUsage);

    return;
  }

  const queues = queueNames
    ? context.getQueues(queueNames)
    : context.getAllQueues();

  const jobCounts = await Promise.all(
    queues.map(async (q) => ({
      queueName: q.name,
      count: await q.getJobCounts(),
    }))
  );

  console.log(jobCounts);
};
