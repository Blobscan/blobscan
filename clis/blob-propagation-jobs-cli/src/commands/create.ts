import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import {
  FINALIZER_WORKER_NAME,
  createBlobPropagationFlowJob,
} from "@blobscan/blob-propagator";
import { prisma } from "@blobscan/db";

import { context } from "../context-instance";
import { env } from "../env";
import type { Command } from "../utils";
import {
  blobHashOptionDef,
  datePeriodOptionDefs,
  helpOptionDef,
  normalizeBlobHashes,
  normalizeDate,
  normalizeStorageQueueName,
} from "../utils";

// TODO: Need to convert it to number explicity due to ci tests failing
// with the following error:
const PRISMA_BATCH_OPERATIONS_MAX_SIZE = parseInt(
  env.PRISMA_BATCH_OPERATIONS_MAX_SIZE.toString()
);
const createCommandOptDefs: commandLineArgs.OptionDefinition[] = [
  helpOptionDef,
  {
    ...blobHashOptionDef,
    description: "Blob hash of the blobs to create jobs for.",
  },
  {
    name: "storage",
    alias: "s",
    typeLabel: "{underline storage}",
    description:
      "Storage used to propagate the selected blobs. Valid values are {italic google}, {italic postgres} or {italic swarm}.",
    multiple: true,
    type: String,
  },
  {
    ...datePeriodOptionDefs.from,
    description: "Date from which to retrieve blobs to create jobs for.",
  },
  {
    ...datePeriodOptionDefs.to,
    description: "Date to which to retrieve blobs to create jobs for.",
  },
];

export const createCommandUsage = commandLineUsage([
  {
    header: "Create Command",
    content: "Create propagation jobs for blobs.",
  },
  {
    header: "Options",
    optionList: createCommandOptDefs,
  },
]);

async function createBlobPropagationJobsInBatches(
  storageQueueNames: string[],
  blobHashesBatchFetcher: (cursorId?: string) => Promise<{
    blobHashes: string[];
    nextCursorId?: string;
  }>
) {
  let cursorId: string | undefined;

  do {
    const { blobHashes, nextCursorId } = await blobHashesBatchFetcher(cursorId);

    const jobs = blobHashes.map((blobHash) =>
      createBlobPropagationFlowJob(
        FINALIZER_WORKER_NAME,
        storageQueueNames,
        blobHash
      )
    );

    if (jobs.length) {
      await context.getPropagatorFlowProducer().addBulk(jobs);
    }

    cursorId = nextCursorId;
  } while (cursorId);
}

export const create: Command = async function (argv) {
  const {
    blobHash: rawBlobHashes,
    help,
    storage: rawStorageNames,
    from: rawFrom,
    to: rawTo,
  } = commandLineArgs(createCommandOptDefs, {
    argv,
  }) as {
    blobHash?: string[];
    help?: boolean;
    storage?: string[];
    from?: string;
    to?: string;
  };

  if (help) {
    console.log(createCommandUsage);

    return;
  }

  const from = rawFrom ? normalizeDate(rawFrom) : undefined;
  const to = rawTo ? normalizeDate(rawTo) : undefined;
  const argStorageNames = rawStorageNames?.map((rawName) =>
    normalizeStorageQueueName(rawName)
  );
  const storageQueues = argStorageNames
    ? context.getQueuesOrThrow(argStorageNames)
    : context.getAllStorageQueues();
  const storageQueueNames = storageQueues.map((q) => q.name);
  let blobHashes: string[];

  if (rawBlobHashes?.length) {
    blobHashes = normalizeBlobHashes(rawBlobHashes);

    const dbBlobs = await prisma.blob.findMany({
      where: {
        versionedHash: {
          in: blobHashes,
        },
      },
    });

    if (dbBlobs.length !== blobHashes.length) {
      const notFoundBlobHashes = blobHashes.filter(
        (hash) => !dbBlobs.find((b) => b.versionedHash === hash)
      );

      throw new Error(
        `Could not find blobs with the following hashes: ${notFoundBlobHashes.join(
          ", "
        )}`
      );
    }

    const jobs = blobHashes.map((blobHash) =>
      createBlobPropagationFlowJob(
        FINALIZER_WORKER_NAME,
        storageQueueNames,
        blobHash
      )
    );

    await context.getPropagatorFlowProducer().addBulk(jobs);
  } else if (from || to) {
    await createBlobPropagationJobsInBatches(
      storageQueueNames,
      async (cursorId) => {
        const dbBlocks = await prisma.block.findMany({
          take: PRISMA_BATCH_OPERATIONS_MAX_SIZE,
          skip: cursorId ? 1 : undefined,
          cursor: cursorId
            ? {
                hash: cursorId,
              }
            : undefined,
          select: {
            hash: true,
            transactions: {
              select: {
                blobs: {
                  select: {
                    blobHash: true,
                  },
                },
              },
            },
          },
          where: {
            timestamp: {
              gte: from,
              lte: to,
            },
          },
        });

        const blobHashes = [
          ...new Set(
            dbBlocks
              .map((b) =>
                b.transactions.map((t) => t.blobs.map((bl) => bl.blobHash))
              )
              .flat(2)
          ),
        ];

        return {
          blobHashes,
          nextCursorId: dbBlocks[dbBlocks.length - 1]?.hash,
        };
      }
    );
  } else {
    await createBlobPropagationJobsInBatches(
      storageQueueNames,
      async (cursorId) => {
        const dbBlobs = await prisma.blob.findMany({
          take: PRISMA_BATCH_OPERATIONS_MAX_SIZE,
          cursor: cursorId
            ? {
                commitment: cursorId,
              }
            : undefined,
          skip: cursorId ? 1 : undefined,
        });
        const blobHashes = dbBlobs.map((b) => b.versionedHash);

        return {
          blobHashes,
          nextCursorId: dbBlobs[dbBlobs.length - 1]?.commitment,
        };
      }
    );
  }
};
