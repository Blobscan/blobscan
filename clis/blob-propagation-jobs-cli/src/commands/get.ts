import type { JobState } from "bullmq";
import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import { buildJobId } from "@blobscan/blob-propagator";
import type { BlobPropagationJob } from "@blobscan/blob-propagator";
import dayjs from "@blobscan/dayjs";
import { prisma } from "@blobscan/db";

import type { QueueHumanName } from "../Context";
import { context } from "../context-instance";
import type { Command } from "../types";
import {
  blobHashOptionDef,
  helpOptionDef,
  queuesOptionDef,
  datePeriodOptionDefs,
  blockRangeOptionDefs,
  slotRangeOptionDefs,
  sortOptionDefs,
} from "../utils";

type JobStatus = (typeof jobStatus)[number];

const jobStatus = [
  "completed",
  "failed",
  "waiting",
  "waiting-children",
  "delayed",
  "prioritized",
] as const;

type PrintableJob = Pick<
  BlobPropagationJob,
  "id" | "timestamp" | "attemptsMade" | "finishedOn" | "failedReason"
> & { index: number; status: JobState | "unknown" };

function printJob(j: PrintableJob) {
  let jobLine = `Job ${j.index !== undefined ? `#${j.index}` : ""} | Id: ${
    j.id
  } | Status: ${j.status} | Timestamp: ${dayjs(j.timestamp)} | Attempts: ${
    j.attemptsMade
  }`;

  if (j.finishedOn) {
    jobLine += ` | Finished Date: ${dayjs(j.finishedOn)}`;
  }

  if (j.failedReason) {
    jobLine += ` | Failed Reason: "${j.failedReason}"`;
  }

  console.log(jobLine);
}

const getCommandOptDefs: commandLineArgs.OptionDefinition[] = [
  helpOptionDef,
  {
    ...blobHashOptionDef,
    description: "Hash of the blob to get the job of",
  },
  {
    ...queuesOptionDef,
    description: `Queue to get the jobs from. ${queuesOptionDef.description}`,
  },
  {
    ...datePeriodOptionDefs.from,
    description: "Date from which to retrieve blobs to get the jobs of.",
  },
  {
    ...datePeriodOptionDefs.to,
    description: "Date to which to retrieve blobs to get the jobs of.",
  },
  {
    ...slotRangeOptionDefs.from,
    description: "Slot from which to retrieve blobs to get the jobs of.",
  },
  {
    ...slotRangeOptionDefs.to,
    description: "Slot to which to retrieve blobs to get the jobs of.",
  },
  {
    ...blockRangeOptionDefs.from,
    description: "Block from which to retrieve blobs to get the jobs of.",
  },
  {
    ...blockRangeOptionDefs.to,
    description: "Block to which to retrieve blobs to get the jobs of.",
  },
  {
    name: "status",
    alias: "t",
    description:
      "Status of the jobs to get. Valid values are {italic completed}, {italic failed}, {italic waiting}, {italic waiting-children}, {italic delayed} or {italic prioritized}.",
    type: (value: string): JobStatus => {
      if (!jobStatus.includes(value as JobStatus)) {
        throw new Error(
          `Invalid status value. Valid values are ${jobStatus.join(", ")}.`
        );
      }

      return value as JobStatus;
    },
    defaultValue: "failed",
  },
  {
    name: "fromIndex",
    alias: "i",
    description: "Index of the first job to get.",
    type: Number,
    defaultValue: 0,
  },
  {
    name: "toIndex",
    alias: "f",
    description: "Index of the last job to get.",
    type: Number,
    defaultValue: 500,
  },
  sortOptionDefs,
];

export const getCommandUsage = commandLineUsage([
  {
    header: "Get Command",
    content: "Get propagation jobs.",
  },
  {
    header: "Options",
    optionList: getCommandOptDefs,
  },
]);

export const get: Command = async function (argv) {
  const {
    blobHash: rawBlobHashes,
    help,
    queue: queueNames,
    fromBlock,
    toBlock,
    fromDate,
    toDate,
    fromSlot,
    toSlot,
    fromIndex,
    toIndex,
    status,
    sort,
  } = commandLineArgs(getCommandOptDefs, {
    argv,
  }) as {
    blobHash?: string[];
    help?: boolean;
    queue?: QueueHumanName[];
    fromBlock: number;
    toBlock: number;
    fromDate?: string;
    toDate?: string;
    fromSlot?: number;
    toSlot?: number;
    fromIndex: number;
    toIndex: number;
    status: JobState;
    sort: "asc" | "desc";
  };

  if (help) {
    console.log(getCommandUsage);

    return;
  }

  const queues = queueNames
    ? context.getQueues(queueNames)
    : context.getAllQueues();
  let blobHashes: string[] = rawBlobHashes
    ? [...new Set(rawBlobHashes.map((blobHash) => blobHash))]
    : [];

  const filtersProvided =
    fromBlock || toBlock || fromDate || toDate || fromSlot || toSlot;

  if (fromBlock && toBlock && toBlock < fromBlock) {
    throw new Error("toBlock must be greater than fromBlock.");
  }

  if (fromSlot && toSlot && toSlot < fromSlot) {
    throw new Error("toSlot must be greater than fromSlot.");
  }

  if (filtersProvided) {
    const dbBlobs = await prisma.blobsOnTransactions.findMany({
      take: toIndex - fromIndex,
      skip: fromIndex,
      select: {
        blobHash: true,
        transaction: {
          select: {
            block: {
              select: {
                timestamp: true,
              },
            },
          },
        },
      },
      orderBy: {
        transaction: {
          block: {
            number: sort,
          },
        },
      },
      where: {
        transaction: {
          block: {
            number:
              fromBlock || toBlock
                ? { gte: fromBlock, lte: toBlock }
                : undefined,
            slot:
              fromSlot || toSlot ? { gte: fromSlot, lte: toSlot } : undefined,
            timestamp:
              fromDate || toDate ? { gte: fromDate, lte: toDate } : undefined,
          },
        },
      },
    });

    if (!dbBlobs.length) {
      console.log("No blobs found with the provided filters.");

      return;
    }

    blobHashes = [...new Set(dbBlobs.map((dbBlob) => dbBlob.blobHash))];
  }

  if (blobHashes.length) {
    const jobPromises = queues.flatMap((q) => {
      let index = fromIndex;

      return blobHashes.map<Promise<PrintableJob | undefined>>(async (hash) => {
        const jobId = buildJobId(q.name, hash);

        const [j, status] = await Promise.all([
          q.getJob(jobId),
          q.getJobState(jobId),
        ]);

        return j ? { ...j, status, index: index++ } : undefined;
      });
    });

    const jobs = await Promise.all(jobPromises);
    const existingJobs = jobs.filter((j): j is PrintableJob => !!j);

    if (!existingJobs?.length) {
      console.log("No blob propagation jobs found.");

      return;
    }

    existingJobs.forEach(printJob);

    return;
  }

  const jobs = (
    await Promise.all(
      queues.map((q) =>
        q.getJobs(status, fromIndex, toIndex, sort === "asc").then((jobs) => {
          let index = fromIndex;

          return jobs.map<PrintableJob>((j) => ({
            ...j,
            index: index++,
            status,
          }));
        })
      )
    )
  ).flat();

  jobs.forEach(printJob);
};
