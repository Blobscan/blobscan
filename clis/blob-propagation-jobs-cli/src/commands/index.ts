export * from "./create";
export * from "./remove";
export * from "./retry";
