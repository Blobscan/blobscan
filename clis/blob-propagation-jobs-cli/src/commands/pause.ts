import { Queue } from "bullmq";
import IORedis from "ioredis";

import { env } from "@blobscan/env";

import type { Command } from "../types";

export const pause: Command = async function () {
  const connection = new IORedis(env.REDIS_URI, { maxRetriesPerRequest: null });

  const overallStatsSyncer = new Queue("swarm-worker", {
    connection,
  });

  // console.log(await overallStatsSyncer.isPaused());

  if (await overallStatsSyncer.isPaused()) {
    await overallStatsSyncer.resume();

    console.log("Resuming overall stats syncer");
  } else {
    console.log("here");
    await overallStatsSyncer.pause();

    console.log("Pausing overall stats syncer");
  }

  // await overallStatsSyncer.close().then(async () => {
  //   await connection.quit();
  // });
};
