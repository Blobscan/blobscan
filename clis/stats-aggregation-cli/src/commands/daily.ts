import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import dayjs from "@blobscan/dayjs";
import type { DatePeriod, RawDatePeriod } from "@blobscan/db";
import { normalizeDailyDatePeriod, prisma } from "@blobscan/db";

import { CommandError } from "../error";
import type { Command, Entity } from "../types";
import {
  ALL_ENTITIES,
  commandLog,
  deleteOptionDef,
  helpOptionDef,
} from "../utils";

type Operation = "populate" | "deleteMany" | "deleteAll";

type OperationResult = { entity: Entity; affectedRows: number };
class DailyCommandError extends CommandError {
  constructor(operation: Operation, error: Error | string) {
    const formattedOperation =
      operation === "deleteMany" ? "deletion" : "aggregation";
    const message = error instanceof Error ? error.message : error;

    super(message, {
      command: "daily",
      operation: formattedOperation,
    });
  }
}

const dailyCommandOptDefs: commandLineUsage.OptionDefinition[] = [
  deleteOptionDef,
  helpOptionDef,
  {
    name: "entity",
    alias: "e",
    typeLabel: "{underline type}",
    description:
      "Entity type to aggregate. Valid values are {italic blob}, {italic block} or {italic tx}.",
    type: String,
    multiple: true,
  },
  {
    name: "from",
    alias: "f",
    typeLabel: "{underline date}",
    description: "Start date in ISO 8601 format.",
    type: String,
  },
  {
    name: "to",
    alias: "t",
    typeLabel: "{underline date}",
    description: "End date in ISO 8601 format.",
    type: String,
  },
];

export const dailyCommandUsage = commandLineUsage([
  {
    header: "Daily Command",
    content: "Aggregate daily stats.",
  },
  {
    header: "Options",
    optionList: dailyCommandOptDefs,
  },
]);

function buildCommandResultMsg(
  operation: Operation,
  datePeriod: DatePeriod,
  opResults: OperationResult[]
) {
  const formattedOperation =
    operation === "deleteMany" ? "deleted" : "calculated";
  const { from, to } = datePeriod || {};
  const formattedFrom = from ? dayjs(from).format("YYYY/MM/DD") : undefined;
  const formattedTo = to ? dayjs(to).format("YYYY/MM/DD") : undefined;
  let period = "";

  if (!from && to) {
    period = `up to ${formattedTo}`;
  } else if (from && !to) {
    period = `from ${formattedFrom}`;
  } else if (from && to) {
    period = `from ${formattedFrom} to ${formattedTo}`;
  }

  const formattedResults =
    opResults
      ?.map((r) => `${r.entity} (${r.affectedRows} rows affected)`)
      .join(", ") || "";

  return `stats ${formattedOperation} successfully ${period} for entities: ${formattedResults}`;
}

async function performDailyStatsOperation(
  entity: Entity,
  operation: Operation,
  datePeriod?: DatePeriod
) {
  let dailyStatsFn;
  const { from, to } = datePeriod || {};
  const isDateProvided = from || to;
  const operation_: Operation =
    operation === "deleteMany" && !isDateProvided ? "deleteAll" : operation;
  const operationParam =
    operation_ === "populate"
      ? datePeriod
      : isDateProvided
      ? {
          where: {
            day: {
              ...(from ? { gte: from } : {}),
              ...(to ? { lte: to } : {}),
            },
          },
        }
      : undefined;

  switch (entity) {
    case "blob":
      dailyStatsFn = prisma.blobDailyStats[operation_];
      break;
    case "block":
      dailyStatsFn = prisma.blockDailyStats[operation_];
      break;
    case "tx":
      dailyStatsFn = prisma.transactionDailyStats[operation_];
      break;
  }

  let result;

  try {
    result = await dailyStatsFn(operationParam).then((res) => {
      return typeof res === "number" ? res : res.count;
    });
  } catch (err) {
    const err_ = err as Error;

    throw new DailyCommandError(
      operation,
      `Failed to perform operation for entity ${entity}: ${err_.message}`
    );
  }

  return result;
}

const daily: Command = async function daily(argv) {
  const {
    from: rawFrom,
    to: rawTo,
    entity: entities,
    // `delete` is a reserved keyword
    delete: delete_,
    help,
  } = commandLineArgs(dailyCommandOptDefs, {
    argv,
  }) as {
    help: boolean;
    delete?: boolean;
    from?: string;
    to?: string;
    entity?: Entity[];
  };

  if (help) {
    console.log(dailyCommandUsage);

    return;
  }

  const operation: Operation = delete_ ? "deleteMany" : "populate";
  const rawDatePeriod: RawDatePeriod = { from: rawFrom, to: rawTo };
  let datePeriod: DatePeriod;
  try {
    datePeriod = normalizeDailyDatePeriod(rawDatePeriod);
  } catch (err) {
    const err_ = err as Error;

    throw new DailyCommandError(operation, err_);
  }
  const selectedEntities = entities?.length ? entities : ALL_ENTITIES;

  const operationResults: OperationResult[] = await Promise.all(
    selectedEntities.map((e) =>
      performDailyStatsOperation(e, operation, datePeriod).then(
        (affectedRows) => ({
          entity: e,
          affectedRows,
        })
      )
    )
  );

  const msg = buildCommandResultMsg(operation, datePeriod, operationResults);
  commandLog("info", "daily", operation, msg);
};

export { daily };
