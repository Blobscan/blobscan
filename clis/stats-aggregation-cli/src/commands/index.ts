export * from "./daily";
export * from "./overall";
