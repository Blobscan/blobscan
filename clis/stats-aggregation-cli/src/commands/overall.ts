import commandLineArgs from "command-line-args";
import commandLineUsage from "command-line-usage";

import type { BlockNumberRange, Prisma } from "@blobscan/db";
import { prisma } from "@blobscan/db";

import { CommandError } from "../error";
import type { Command } from "../types";
import { commandLog, deleteOptionDef, helpOptionDef } from "../utils";

type BlockId = number | "latest" | "finalized";

const PRISMA_BLOCKS_BATCH_SIZE = 2_000_000;

class OverallCommandError extends CommandError {
  constructor(operation: string, error: Error | string) {
    super(error instanceof Error ? error.message : error, {
      command: "overall",
      operation,
    });
  }
}

const overallCommandOptDefs: commandLineUsage.OptionDefinition[] = [
  deleteOptionDef,
  helpOptionDef,
  {
    name: "to",
    alias: "t",
    typeLabel: "{underline block-id}",
    description: `Block identifier up to which to aggregate data. It can be a block number, "latest" for the last indexed block or "finalized"  for the chain's most recent finalized block. It defaults to "finalized"`,
    defaultValue: "finalized",
    type: String,
  },
  {
    name: "batchSize",
    alias: "s",
    typeLabel: "{underline size}",
    description: `Number of blocks to process in a single batch. It defaults to ${PRISMA_BLOCKS_BATCH_SIZE}`,
    defaultValue: PRISMA_BLOCKS_BATCH_SIZE,
    type: Number,
  },
];

export const overallCommandUsage = commandLineUsage([
  {
    header: "Overall Command",
    content: "Aggregate overall stats.",
  },
  {
    header: "Options",
    optionList: overallCommandOptDefs,
  },
]);

async function resolveBlockId(blockId: BlockId): Promise<number> {
  try {
    if (typeof blockId === "number" || !isNaN(Number(blockId))) {
      return Number(blockId);
    } else if (blockId === "finalized") {
      const syncState = await prisma.blockchainSyncState.findUnique({
        select: {
          lastFinalizedBlock: true,
        },
        where: {
          id: 1,
        },
      });

      return syncState?.lastFinalizedBlock ?? 0;
    } else {
      return (await prisma.block.findLatest().then((b) => b?.number)) ?? 0;
    }
  } catch (err) {
    const err_ = err as Error;

    throw new OverallCommandError("aggregation", err_);
  }
}

async function deleteOverallStats() {
  const newBlockchainSyncState: Partial<Prisma.BlockchainSyncStateGroupByOutputType> =
    {
      lastAggregatedBlock: null,
    };

  await prisma.$transaction([
    prisma.blobOverallStats.deleteMany(),
    prisma.blockOverallStats.deleteMany(),
    prisma.transactionOverallStats.deleteMany(),
    prisma.blockchainSyncState.upsert({
      create: newBlockchainSyncState,
      update: newBlockchainSyncState,
      where: {
        id: 1,
      },
    }),
  ]);
}

async function incrementOverallStats(
  blockRange: BlockNumberRange,
  {
    batchSize = PRISMA_BLOCKS_BATCH_SIZE,
  }: {
    batchSize?: number;
  }
) {
  const { from, to } = blockRange;
  const unprocessedBlocks = to - from + 1;
  const batches = Math.ceil(unprocessedBlocks / batchSize);

  for (let i = 0; i < batches; i++) {
    const batchFrom = from + i * batchSize;
    const batchTo = Math.min(batchFrom + batchSize - 1, to);
    const blockRange: BlockNumberRange = { from: batchFrom, to: batchTo };
    const newBlockchainSyncState: Partial<Prisma.BlockchainSyncStateGroupByOutputType> =
      {
        lastAggregatedBlock: batchTo,
      };

    await prisma.$transaction([
      prisma.blockOverallStats.increment(blockRange),
      prisma.transactionOverallStats.increment(blockRange),
      prisma.blobOverallStats.increment(blockRange),
      prisma.blockchainSyncState.upsert({
        create: newBlockchainSyncState,
        update: newBlockchainSyncState,
        where: {
          id: 1,
        },
      }),
    ]);

    if (batches > 1) {
      const formattedBlockRange = `${blockRange.from.toLocaleString()}-${blockRange.to.toLocaleString()}`;
      const batch = batches > 1 ? `(batch ${i + 1}/${batches})` : "";

      commandLog(
        "debug",
        "overall",
        "increment",
        `stats calculated for block range ${formattedBlockRange} (${batch})`
      );
    }
  }
}

const overall: Command = async function (argv) {
  const {
    batchSize,
    to,
    delete: delete_,
    help,
  } = commandLineArgs(overallCommandOptDefs, {
    argv,
  }) as {
    batchSize: number;
    to?: string;
    delete?: boolean;
    help: boolean;
  };

  if (help) {
    console.log(overallCommandUsage);

    return;
  }

  if (delete_) {
    await deleteOverallStats();

    commandLog(
      "info",
      "overall",
      "deleteMany",
      `All overall stats removed successfully`
    );

    return;
  }

  let targetBlockId: BlockId;

  if (!isNaN(Number(to))) {
    targetBlockId = Number(to);
  } else if (to === "latest" || to === "finalized") {
    targetBlockId = to;
  } else {
    throw new OverallCommandError(
      "aggregation",
      `Invalid \`to\` flag value. Expected a block number, "latest" or "finalized" but got ${to}`
    );
  }
  let targetBlockNumber = await resolveBlockId(targetBlockId);
  const [lastSyncedFinalizedBlock, latestIndexedBlock] = await Promise.all([
    prisma.blockchainSyncState
      .findUnique({
        where: {
          id: 1,
        },
      })
      .then((state) => state?.lastAggregatedBlock ?? 0),
    prisma.block.findLatest().then((b) => b?.number),
  ]);

  // If we haven't indexed any blocks yet, don't do anything
  if (!latestIndexedBlock) {
    commandLog(
      "info",
      "overall",
      "increment",
      "Skipping as there as no blocks have been indexed yet"
    );

    return;
  }

  const fromBlockNumber = lastSyncedFinalizedBlock + 1;

  // Process only finalized blocks that were indexed
  if (targetBlockNumber > latestIndexedBlock) {
    targetBlockNumber = latestIndexedBlock;
  }

  if (fromBlockNumber > targetBlockNumber) {
    commandLog(
      "info",
      "overall",
      "increment",
      `Skipping as there is no new finalized blocks (Last processed finalized block: ${lastSyncedFinalizedBlock.toLocaleString()})`
    );

    return;
  }

  await incrementOverallStats(
    { from: fromBlockNumber, to: targetBlockNumber },
    { batchSize }
  );

  commandLog(
    "info",
    "overall",
    "increment",
    `All stats incremented successfully from block ${fromBlockNumber.toLocaleString()} to ${targetBlockNumber.toLocaleString()}`
  );

  return;
};

export { overall };
