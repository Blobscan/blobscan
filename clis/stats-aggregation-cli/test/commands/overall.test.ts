import { afterAll, beforeAll, describe, expect, it, vi } from "vitest";

import { prisma } from "@blobscan/db";
import { omitDBTimestampFields } from "@blobscan/test";

import { overall, overallCommandUsage } from "../../src/commands/overall";
import { runHelpArgTests } from "../helpers";

function getAllOverallStats() {
  return Promise.all([
    prisma.blobOverallStats.findMany(),
    prisma.blockOverallStats.findMany(),
    prisma.transactionOverallStats.findMany(),
  ]);
}

describe("Overall command", () => {
  beforeAll(() => {
    // Silence console.log
    vi.spyOn(console, "log").mockImplementation(() => void {});
  });

  afterAll(() => {
    vi.clearAllMocks();
  });

  runHelpArgTests(overall, overallCommandUsage);

  describe("when incrementing overall stats", () => {
    it("should increment them until latest indexed block when target block 'latest' is provided", async () => {
      await overall(["--to", "latest"]);

      const overallStats = await getAllOverallStats().then((overallStats) =>
        overallStats.map((stats) => stats.map(omitDBTimestampFields))
      );

      expect(overallStats).toMatchSnapshot();
    });

    it("should increment them until latest finalized block", async () => {
      await overall(["--to", "finalized"]);

      const overallStats = await getAllOverallStats().then((overallStats) =>
        overallStats.map((stats) => stats.map(omitDBTimestampFields))
      );

      expect(overallStats).toMatchSnapshot();
    });

    it("should increment until provided block number", async () => {
      await overall(["--to", "1005"]);

      const overallStats = await getAllOverallStats().then((allOverallStats) =>
        allOverallStats.map((stats) => stats.map(omitDBTimestampFields))
      );

      expect(overallStats).toMatchInlineSnapshot(`
        [
          [
            {
              "avgBlobSize": 1000,
              "id": 1,
              "totalBlobSize": 4000n,
              "totalBlobs": 4,
              "totalUniqueBlobs": 0,
            },
          ],
          [
            {
              "avgBlobAsCalldataFee": 5000000,
              "avgBlobFee": 100000000,
              "avgBlobGasPrice": 20,
              "id": 1,
              "totalBlobAsCalldataFee": "5000000",
              "totalBlobAsCalldataGasUsed": "250000",
              "totalBlobFee": "100000000",
              "totalBlobGasUsed": "5000000",
              "totalBlocks": 1,
            },
          ],
          [
            {
              "avgMaxBlobGasFee": 100,
              "id": 1,
              "totalTransactions": 2,
              "totalUniqueReceivers": 0,
              "totalUniqueSenders": 0,
            },
          ],
        ]
      `);
    });
  });

  it("should fail when an invalid target block height is given", async () => {
    expect(
      overall(["--to", "invalid-block-height"])
    ).rejects.toMatchInlineSnapshot(
      '[Error: Overall stats aggregation failed: Invalid `to` flag value. Expected a block number, "latest" or "finalized" but got invalid-block-height]'
    );
  });
});
