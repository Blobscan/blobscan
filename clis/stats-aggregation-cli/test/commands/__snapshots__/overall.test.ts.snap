// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Overall command > when incrementing overall stats > should increment them until latest finalized block 1`] = `
[
  [
    {
      "avgBlobSize": 1175,
      "id": 1,
      "totalBlobSize": 9400n,
      "totalBlobs": 8,
      "totalUniqueBlobs": 1,
    },
  ],
  [
    {
      "avgBlobAsCalldataFee": 5406666.666666667,
      "avgBlobFee": 114000000,
      "avgBlobGasPrice": 21.33333333333333,
      "id": 1,
      "totalBlobAsCalldataFee": "16220000",
      "totalBlobAsCalldataGasUsed": "760000",
      "totalBlobFee": "342000000",
      "totalBlobGasUsed": "16000000",
      "totalBlocks": 3,
    },
  ],
  [
    {
      "avgMaxBlobGasFee": 100,
      "id": 1,
      "totalTransactions": 4,
      "totalUniqueReceivers": 0,
      "totalUniqueSenders": 0,
    },
  ],
]
`;

exports[`Overall command > when incrementing overall stats > should increment them until latest indexed block when target block 'latest' is provided 1`] = `
[
  [
    {
      "avgBlobSize": 1155.555555555556,
      "id": 1,
      "totalBlobSize": 10400n,
      "totalBlobs": 9,
      "totalUniqueBlobs": 1,
    },
  ],
  [
    {
      "avgBlobAsCalldataFee": 5457500,
      "avgBlobFee": 115750000,
      "avgBlobGasPrice": 21.5,
      "id": 1,
      "totalBlobAsCalldataFee": "21830000",
      "totalBlobAsCalldataGasUsed": "1015000",
      "totalBlobFee": "463000000",
      "totalBlobGasUsed": "21500000",
      "totalBlocks": 4,
    },
  ],
  [
    {
      "avgMaxBlobGasFee": 100,
      "id": 1,
      "totalTransactions": 5,
      "totalUniqueReceivers": 0,
      "totalUniqueSenders": 0,
    },
  ],
]
`;
